### ISSUE TITLE
wxStaticBoxSizer subclassing

### ISSUE THREAD

**Issue by FillipMatthew** [wxFormBuilder/wxFormBuilder#822]
wxStaticBoxSizer supports passing in a pointer to a wxStaticBox but wxFormBuilder does not yet support this.

Would it be possible to add support for one of the following?:
1. Subclassing of sizers.
2. More control over sizer constructor parameters. This though might only work for cases like wxStaticBoxSizer.

**Comment by sodevel** [comment]
Sizers are an essential part of wxFB layouts and sadly require hardcoded things beside their pure XML definition. Right now i don't know how complicated it would be to change something sizer related. However, im pretty sure subclassing would be quite a lot of work, but i don't think this is necessary. Adding an additional parameter should work and should be easier than subclassing, however, there is the problem, wxFB does not model a wxStaticBox. So where should that wxStaticBox come from? You could use the CustomCode component to inject code to create one but thats not very convenient.

Actually, how do you use a wxStaticBox without a wxStaticBoxSizer? In my 10+ years of using wxWidgets i never used a wxStaticBox on its own, i always used a wxStaticBoxSizer.

**Comment by FillipMatthew** [comment]
I don't know of any use for wxStaticBox without wxStaticBoxSizer either.

Currently I'm using the custom control (I didn't know how to use the CustomCode component). I manually edit the generated code after. This required removing the line that adds my control to the parent sizer, and replacing the new wxStaticBox with my instance. It also leaves the control in wxFB meaning the preview isn't accurate.

If the wxStaticBoxSizer has a field to override the construction of the wxStaticBox that is currently used, that would work, but may also need a second field for the header in case someone wanted to use pass a subclass instead like my case. This would allow writing out something like 'new MyStaticBox()', or passing in an instance created by the custom control or CustomCode component.

**Comment by sodevel** [comment]
Sorry for the late reply, but i had been quite busy with work.

So basically all you want to do is pass in a custom wxStaticBox for the wxStaticBoxSizer? I checked the code and it looks like this is possible to do without too many changes. However, you won't be able to see this in the preview, beside other reasons, this would require that wxFB can compile and use the code of your wxStaticBox implementation at runtime and this is not possible. If you want to have it in the preview, you might be able to write your own plugin that creates such a widget, but the plugin system is quite limited, you cannot add completely new widget types without changing the core.

So what i could do is to add a new property to the wxStaticBoxSizer component that allows you to pass in the name of a wxStaticBox element so that it doesn't create one by itself. To actually create that element, you have to use the CustomCode component. The component allows you to inject arbitrary code into the generated code roughly at the position where it appears in the object tree. This is almost the same what the CustomControl component does with one significant difference: The CustomControl element is handled by wxFB in its data model like a wxWindow component so e.g. it gets added to the containing sizer. The CustomCode component however is completely invisible in the data model, this allows you to freely inject code into the generated code.

**Comment by FillipMatthew** [comment]
No worries, thanks for responding.

Exactly, that would work perfectly. Would be nice to be able to specify a member variable for the custom code. That isn't the end of the world since I can still use the GetStaticBox() with a reinterpret_cast.

I'm not worried about the preview, not having to edit the code every time after generating the base files helps a tone. Thank you.

**Comment by sodevel** [comment]
I have added a new property for wxStaticBoxSizer (plus some other changes) in the branch [feature/custom-staticbox](https://github.com/wxFormBuilder/wxFormBuilder/tree/feature/custom-staticbox) that should fit your needs. If you don't want/can't compile wxFB yourself you can grab a build from GitHub Actions.

**Comment by FillipMatthew** [comment]
Thanks @sodevel, changes look good. I haven't compiled anything with it yet, but I did noticed a few things.

'myStaticBox->SetLabel( _("label") )' is still generated, even though it will probably already be set when writing the code for the CustomCode component. I also don't know if this will cause problems if I've already passed a wxCheckBox as the label.

The custom code component can't be a child of other code, so it can only be at the begining or end of a wxDialog constructor for example. I also suspect that if this changes it will still need special changes to be directly on a wxPanel. Since if you have a wxStaticBoxSizer on a wxPanel, you need to parent the wxStaticBox to the same panel as the sizer is on? When I had tried the wxCustomControl instead I also had to create the extra sizer to wrap the wxStaticBoxSizer otherwise I couldn't get my code to have the same parent.

**Comment by sodevel** [comment]
> 'myStaticBox->SetLabel( _("label") )' is still generated

This is by intent, i wanted to use as much data that is defined in wxFB. Looking at the `wxStaticBoxSizer` constructors however, it seems to make sense to not use the label for a custom wxStaticBox, the label is only used when the wxStaticBox is created by the sizer itself. I will change that.

> The custom code component can't be a child of other code, so it can only be at the begining or end of a wxDialog constructor for example.

Indeed, this is because this component has an own type and currently this type can only be a direct child of a form. This limitation was fine for my uses cases so far, but i think this can be relaxed. Sadly there are a lot of types defined and i do not want to mess up all the type definitions for this artifical component. Maybe it is enough to add it to some types so that you have more injection points for the most common use cases. I need to check if this messes up code generation somewhere, and also if this has some bad side effects for the only other component of that type.

> I also suspect that if this changes it will still need special changes to be directly on a wxPanel. Since if you have a wxStaticBoxSizer on a wxPanel, you need to parent the wxStaticBox to the same panel as the sizer is on?

The CustomCode component is "invisible" in the data model, its code fragment gets inserted as-is, the position only controls where in the generated code this fragment gets inserted. So you have to setup the correct relationships in the code fragment yourself.

> When I had tried the wxCustomControl instead I also had to create the extra sizer to wrap the wxStaticBoxSizer otherwise I couldn't get my code to have the same parent.

The CustomControl component is a wxWindow in the data model and treated like that, so yes, you might need to setup some structure to get the generated code right. Therefor this component might not be the best choice for your use case.

**Comment by FillipMatthew** [comment]
Maybe this isn't useful now but I had an idea about positioning the code for the CustomCode component. If the component had a property to specify which control to generate the code after or before, then you wouldn't need to place it on any control in wxFB it just needs to specify where to generate the code. It could be a text field for a name or a dropdown list of controls. Maybe that just over complicates generation though.

**Comment by sodevel** [comment]
Yes, this is very complicated and a lot of work for little benefit. Right now the code generator basically walks through the object tree and calls the code generator templates for each object and simply outputs the result, there is no way to navigate through the output and specify insertion points. I did some tests in the branch https://github.com/wxFormBuilder/wxFormBuilder/tree/feature/nonvisual-placement, this allows to use CustomCode components as sizer child items, this adds a lot of insertion points with very little work, maybe you want to try that one.

Right now i'm not going to merge these branches into master because for them it is required to update the project file version and i don't want to activate this right now.

**Comment by FillipMatthew** [comment]
Thanks, these changes will work well together. I can work with this nicely now.

Maybe in the future when there is more time for other changes the custom code could also be placed on panels. At the moment you can't place a custom static box sizer as the first child of a panel. A simple workaround is to just place the custom static box and sizer inside a box sizer but it would be nice to not need that. Should I bother opening a separate issue for this since my original issue is solved?
